### PR TITLE
Allow OTHER type

### PR DESCRIPTION
**Pull request by Sammers21**
So any types will be allowed to use. Then different languages should adopt it.

### CODE REVIEW

**Comment by Sammers21** [comment]
@vietj , can you review?

**Comment by vietj** [comment]
I think we need something more fine grained that just this that removes all check.

I think what we want to change is to allow API_TYPE == OTHER_TYPE is argument and return declarations.

**Comment by Sammers21** [comment]
@vietj , i have added **codegen.allow.any.types** option. By default it is false.

> need to actually check at more fine grained level when we check that a type is API, we should change to (API || OTHER)

For now what we have is just two levels of check: **API** and **OTHER**, so either we allow OTHER, or not. Why do we need to introduce check levels when we can stick with just true/false property?

**Comment by vietj** [comment]
actually in my mail I said we don't want to use a codegen param because it will not work in the main aggregator, can you revert the change ?

**Comment by vietj** [comment]
concerning the check, it should not be done this way but instead done where the method `isVertxGenInterface` is used.

wherever we have this method used we should also have `other` return true, e.g:

```
  private boolean isLegalNonCallableReturnType(TypeInfo type) {
    if (type.getKind().basic) {
      return true;
    }
    if (type.getKind().json) {
      return true;
    }
    if (isLegalDataObjectTypeReturn(type)) {
      return true;
    }
    if (isLegalEnum(type)) {
      return true;
    }
    if (type.getKind() == ClassKind.THROWABLE) {
      return true;
    }
    if (isTypeVariable(type)) {
      return true;
    }
    if (type.getKind() == ClassKind.OBJECT) {
      return true;
    }
    if (isVertxGenInterface(type, true)) {
      return true;
    }
    if (type.getKind() == ClassKind.OTHER) {
      return true;
    }
    if (isLegalContainerReturn(type)) {
      return true;
    }
    return false;
  }
```

there are 4 places where it should be changed

**Comment by Sammers21** [comment]
@vietj, you mean that `isVertxGenInterface` should have a boolean param, like `allowAnyType`? So the check can be done inside the method?

**Comment by Sammers21** [comment]
@vietj, got it. But I think having it in just two places is fine

**Comment by vietj** [comment]
we want to handle method parameters and method return types

also we need you to add tests for this behavior as well as an extra (optional) TCK class for testing this

**Comment by Sammers21** [comment]
@vietj , yes, you are right.

**Comment by Sammers21** [comment]
When i runned codegen(with this brachn changes) on cassandra client project i m getting:

```
SEVERE: io.vertx.codegen.CodeGenProcessor - Could not generate model for io.vertx.cassandra.CassandraClient: No type mapping found for param type OTHER.com.datastax.driver.core.Statement
io.vertx.codegen.GenException: No type mapping found for param type OTHER.com.datastax.driver.core.Statement
	at io.vertx.codegen.ClassModel.traverseType(ClassModel.java:693)
	at io.vertx.codegen.ClassModel.process(ClassModel.java:553)
	at io.vertx.codegen.CodeGen$ModelEntry.getValue(CodeGen.java:188)
	at io.vertx.codegen.CodeGen$ModelEntry.getValue(CodeGen.java:168)
	at io.vertx.codegen.CodeGenProcessor.lambda$process$6(CodeGenProcessor.java:145)
	at java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1625)
	at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:743)
	at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:742)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
	at io.vertx.codegen.CodeGenProcessor.process(CodeGenProcessor.java:143)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
	at com.sun.tools.javac.main.Main.compile(Main.java:523)
	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
	at org.bsc.maven.plugin.processor.AbstractAnnotationProcessorMojo.executeWithExceptionsHandled(AbstractAnnotationProcessorMojo.java:650)
	at org.bsc.maven.plugin.processor.AbstractAnnotationProcessorMojo.execute(AbstractAnnotationProcessorMojo.java:379)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:956)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:290)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:194)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)

[ERROR] diagnostic: /Users/sammers/vertx-cassandra-client/src/main/java/io/vertx/cassandra/CassandraClient.java:35: error: Could not generate model for io.vertx.cassandra.CassandraClient: No type mapping found for param type OTHER.com.datastax.driver.core.Statement
public interface CassandraClient {
```

**Comment by Sammers21** [comment]
It seems that I need to add two entries to the file https://github.com/vert-x3/vertx-codegen/blob/master/src/main/resources/lang-type-mapping.properties

**JS.OTHER** and **GROOVY.OTHER** properties

**Comment by Sammers21** [comment]
@vietj , what do you think?

**Comment by vietj** [comment]
I think that actually when the overload type checker should not process method that have OTHER type because they won't be translated

**Comment by Sammers21** [comment]
What is overload checking?

**Comment by vietj** [comment]
it's the `MethodOverloadChecker` class that threw this exception

we don't want to use it for OTHEr methods

**Comment by Sammers21** [comment]
@vietj, about tests. Would not it be fine just to change broken tests, so they will cover this PR introduced functionality as I just did?

**Comment by vietj** [comment]
I'm not sure what you mean. I think we need to change existing tests (for new behavior) and also have new test for testing these new methods.

**Comment by Sammers21** [comment]
Yes, that is what I mean, I changed existing tests for new behaviour and I think that we don't need new tests because changed ones cover new behaviour. Can you review?a

**Comment by vietj** [comment]
at least for each `ClassKind.API` we need the same test with `ClassKind.OTHER`

**Comment by vietj** [comment]
In addition I think we need also to cover the case container param (that is not covered by the API type) to provide this more flexiblity (and thus we need test for this)

**Comment by Sammers21** [comment]
> at least for each ClassKind.API we need the same test with ClassKind.OTHER

What exaclty you mean by this? Do I need to grep test sources, then to find test cases where _ClassKind.API_ is used, and try to create a similar test, but for `ClassKind.OTHER`?

> In addition I think we need also to cover the case container param (that is not covered by the API type) to provide this more flexiblity (and thus we need test for this)

I have just added tests for this. Can you look at it?

**Comment by Sammers21** [comment]
I realized that container type does not become other now. Where the code responsible for transformation `java.util.List -> ClassKind.???` ?

**Comment by vietj** [comment]
it is in `ClassKind` as a static method

**Comment by vietj** [comment]
but I think it's not the right place to do it

**Comment by vietj** [comment]
@Sammers21 proxy stuff is : https://github.com/vert-x3/vertx-service-proxy/blob/master/src/main/java/io/vertx/serviceproxy/model/

**Comment by Sammers21** [comment]
@vietj , I wonder if `Map<OTHER_TYPE, OTHER_TYPE>` should be allowed

**Comment by Sammers21** [comment]
because for now, the only valid key type for map container is `String`. So maybe it makes any sense to allow `Map<String, OTHER_TYPE>`, not `Map<OTHER_TYPE, OTHER_TYPE>`

**Comment by Sammers21** [comment]
@vietj , can you review one more time, please? I think I covered the new behaviour, but I m not sure.

File: src/test/java/io/vertx/test/codegen/testapi/MethodWithOtherContainerType.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package io.vertx.test.codegen.testapi;
+
+import io.vertx.codegen.annotations.VertxGen;
+
+import java.net.Socket;
+import java.util.Set;
+
+@VertxGen
+public interface MethodWithOtherContainerType {
```
Comment (vietj) [inline src/test/java/io/vertx/test/codegen/testapi/MethodWithOtherContainerType.java]: this should also handle map and list

Comment (Sammers21) [reply src/test/java/io/vertx/test/codegen/testapi/MethodWithOtherContainerType.java]: Just added

**Comment by Sammers21** [comment]
>I think we'll handle container type promotion to other separately

Not sure that I got what you mean by this. 

Can you review one more time, please?

**Comment by vietj** [comment]
I mean the promotion of Map<String, Other> to OTHER kind

**Comment by vietj** [comment]
we also need a new TCK type to check the other types (for instance name it `EnhancedMethodsTCK`)

**Comment by vietj** [comment]
and this TCK type will be used by vertx-rx to verify the behavior

**Comment by Sammers21** [comment]
I am not sure what is TCK. 

I see there is a dir: **src/tck**. And here I can find classes/interfaces, whose name ended with TCK. Then I looked at `CollectionTCK` interface. In this interface, there is a lot of different methods, which has return type/params type related to collections.

Do i need to create a similar thing? So, to create an interface(`EnhancedMethodsTCK`)  and declare here all possible combinations of signatures allowed by the new behavoir?

**Comment by Sammers21** [comment]
@vietj, just added `EnhancedMethodsTCK`, can you review?

**Comment by Sammers21** [comment]
> can the TCK method actually do something?

You mean that implementation should do something? I thought now it is doing something.

**Comment by Sammers21** [comment]
>can you also use this TCK in rx-java and provide a PR, so we can review both at the same time and see how it goes ?

Do I need to generate rx-fied API for `EnhancedMethodsTCK` and create some tests with it?

**Comment by vietj** [comment]
yes that's what I mean.

the TCK is used by generators to check the implementation is valid.

e.g `io.vertx.rx.java.test.gen.FunctionParamTCKTest`

in vertx-rx-java

etc...

**Comment by Sammers21** [comment]
@vietj, here is the PR: https://github.com/vert-x3/vertx-rx/pull/150

**Comment by vietj** [comment]
Remove system.out prints

**Comment by Sammers21** [comment]
@vietj, done

**Comment by vietj** [comment]
I would like to have a few tests +  TCK using arrays (and primitive arrays), so

`Object[]`, `Object...`, `String[]`, `int[]`, `byte[]`

at least

**Comment by Sammers21** [comment]
@vietj , is it fine now?

**Comment by vietj** [comment]
I think it's fine for this PR, I need to review the rx PR

**Comment by Sammers21** [comment]
@vietj , does it means that `@SuppressWarnings("codegen-enhanced-method") ` feature should be implemented in a separated PR?

**Comment by vietj** [comment]
@Sammers21 no it should be in the same PR actually

**Comment by vietj** [comment]
I've pushed a rebased and squashed branch https://github.com/vert-x3/vertx-codegen/tree/Sammers21-allow_other

**Comment by Sammers21** [comment]
@vietj, good. Should it be merged into the master branch?

**Comment by vietj** [comment]
@Sammers21 still working on it

**Comment by vietj** [comment]
rebased and merged

**Comment by Sammers21** [comment]
@vietj , great. Thanks!

**Comment by vietj** [comment]
yeah, took some time out to figure out

**Comment by vietj** [comment]
I've changed the name of suppress warning tag to _codegen-allow-any-java-type_ btw and now it's mandatory

**Pull request closed without merging**